psycopg/psycopg/types/composite.py

```diff
@@ -219,63 +223,136 @@ def __init__(self, oid: int, context: abc.AdaptContext | None = None):
         # records), oids might differ and we'd need separate transformers.
         self._txs: dict[tuple[int, ...], abc.Transformer] = {}
 
-    def load(self, data: abc.Buffer) -> tuple[Any, ...]:
+    def _parse_record(self, data: abc.Buffer) -> tuple[Any, ...]:
         nfields = unpack_len(data, 0)[0]
+        if not nfields:
+            return ()
+
+        oids, record = zip(*self._iterunpack(data, nfields))
+        tx = self._get_transformer(oids)
+        return tx.load_sequence(record)
+
+    @staticmethod
+    def _iterunpack(
+        data: abc.Buffer, nfields: int
+    ) -> Iterator[tuple[int, abc.Buffer | None]]:
+        # Iter through the buffer and yield `OID, Buffer` pairs
         offset = 4
-        oids = []
-        record = []
         for _ in range(nfields):
             oid, length = _unpack_oidlen(data, offset)
             offset += 8
-            record.append(data[offset : offset + length] if length != -1 else None)
-            oids.append(oid)
+            field = data[offset : offset + length] if length != -1 else None
             if length >= 0:
                 offset += length
 
-        key = tuple(oids)
-        try:
-            tx = self._txs[key]
-        except KeyError:
-            tx = self._txs[key] = Transformer(self._ctx)
-            tx.set_loader_types(oids, self.format)
+            yield oid, field
+
+    def _get_transformer(self, key: tuple[int, ...]) -> abc.Transformer:
+        # This could be managed with a custom `dict.__missing__` implementation
+        # micro-optimization - Prefer a membership check to a KeyError
+        #   try/except blocks are expensive in hot loops.
+        if key in self._txs:
+            return self._txs[key]
+
+        tx = Transformer(self._ctx)
+        tx.set_loader_types([*key], self.format)
+        self._txs[key] = tx
+        return tx
+
+
+class RecordBinaryLoader(BaseBinaryCompositeLoader):
+    format = pq.Format.BINARY
+
+    def __init__(self, oid: int, context: abc.AdaptContext | None = None):
+        super().__init__(oid, context)
+        self._ctx = context
+        # Cache a transformer for each sequence of oid found.
+        # Usually there will be only one, but if there is more than one
+        # row in the same query (in different columns, or even in different
+        # records), oids might differ and we'd need separate transformers.
+        self._txs: dict[tuple[int, ...], abc.Transformer] = {}
+
+    def load(self, data: abc.Buffer) -> tuple[Any, ...]:
+        return self._parse_record(data)
+
 
-        return tx.load_sequence(tuple(record))
+InstanceT = TypeVar("InstanceT")
```

Comment:
Within the scope of the `register_composite` and class creation, mypy can track it, which will still provide more robust analysis - in this case, it caught some subtle bugs which were otherwise masked by passing through `Any` rather than the anonymous `InstanceT` typevar.